src/lib/matchers.rs

```diff
@@ -216,19 +281,264 @@ impl<T: Clone> MappingStore<T> {
     /// To nodes are common if they have already been matched.
     fn num_common_descendants(&self, from: &NodeId, to: &NodeId) -> u32 {
         let mut dst_desc = HashSet::new();
-        for node in to.breadth_first_traversal(&self.to) {
+        for node in to.breadth_first_traversal(&self.to_arena) {
             dst_desc.insert(node);
         }
         let mut common = 0;
         let mut to: Option<NodeId>;
-        for node in from.descendants(&self.from) {
+        for node in from.descendants(&self.from_arena) {
             to = self.get_to(&node);
             if to.is_some() && dst_desc.contains(&to.unwrap()) {
                 common += 1;
             }
         }
         common
     }
+
+    /// Given a mapping store, generate an edit script.
+    ///
+    /// This function implements the optimal algorithm of Chawathe et al. (1996).
+    /// Variable names as in Figures 8 and 9 of the paper.
+    pub fn generate_edit_script(&mut self, root: NodeId) -> EditScriptResult<T> {
+        let mut script: EditScript<T> = EditScript::new();
+        let mut from_in_order: HashSet<NodeId> = HashSet::new();
+        let mut to_in_order: HashSet<NodeId> = HashSet::new();
+        let mut next_from_id = self.from_arena.size();
+        let mut new_mappings = TemporaryMappingStore::new();
+        // Copy current mappings over to new_mappings.
+        for (key, value) in &self.from {
+            new_mappings.push(*key, value.0);
+        }
+        // Combined update, insert, align and move phases.
+        let tmp_to_arena = self.to_arena.clone();
+        for x in root.breadth_first_traversal(&tmp_to_arena) {
+            if x.is_root(&self.to_arena) {
```

Comment:
That also leads to a compiler error and Rust is suggesting implementing a trait from a nightly feature called `core::slice::SliceExt`, which seems like overkill. However, the bigger issue with this algorithm is that it mutates the same tree it iterates over. In this implementation I have `cloned()` the `to_arena` AST. If that turns out to be an expensive operation when we start benchmarking, it may be better to change the underlying AST library to use `Cell` references or similar.